core/src/main/java/org/elasticsearch/common/ParseField.java

```diff
@@ -30,36 +30,33 @@
 
     private static final DeprecationLogger DEPRECATION_LOGGER = new DeprecationLogger(Loggers.getLogger(ParseField.class));
 
-    private final String camelCaseName;
     private final String underscoreName;
     private final String[] deprecatedNames;
     private String allReplacedWith = null;
 
     public ParseField(String value, String... deprecatedNames) {
-        camelCaseName = Strings.toCamelCase(value);
         underscoreName = Strings.toUnderscoreCase(value);
-        if (deprecatedNames == null || deprecatedNames.length == 0) {
-            this.deprecatedNames = Strings.EMPTY_ARRAY;
-        } else {
-            final HashSet<String> set = new HashSet<>();
-            for (String depName : deprecatedNames) {
-                set.add(Strings.toCamelCase(depName));
-                set.add(Strings.toUnderscoreCase(depName));
-            }
-            this.deprecatedNames = set.toArray(new String[set.size()]);
+        final HashSet<String> set = new HashSet<>();
+        String camelCaseName = Strings.toCamelCase(value);
+        if (camelCaseName.equals(value) == false) {
+            set.add(camelCaseName);
+        }
+        for (String depName : deprecatedNames) {
+            set.add(Strings.toCamelCase(depName));
+            set.add(Strings.toUnderscoreCase(depName));
         }
```

Comment:
Luca's comment makes me wonder if we should also check whether `set` contains `value` and throw an IAE in that case to protect against user errors.